### PR TITLE
Basic Dart client

### PR DESCRIPTION
**Pull request by dmdashenkov**
# Dart client

Introducing a new client for the Spine web API. The client is written in Dart and targets any Dart-based frontend technologies, such as Flutter and Angular Dart.

The client mirrors the structure of Java and JS clients.
The user works with two main components:
 - `BackendClient` — a facade for communications with server-side;
 - `ActorRequestFactory` — a factory of requests handled by Spine, commands and queries.

Uses the Firebase web API to communicate with the server side.
Entity/event subscription is not supported in this version.
### Usage
```dart
void main() async {
    var firebase = RestClient(fb.FirebaseClient.anonymous(), FIREBASE_URL);
    var client = BackendClient(BACKEND_URL, firebase);

    var actor = UserId();
    actor.value = newUuid();
    var requestFactory = ActorRequestFactory(actor);

    var cmd = createCommandMessage();
    var anyCmd = Any.pack(cmd, typeUrlPrefix: 'type.example.org');
    // Post command and wait for acknowledgement.
    await client.post(requestFactory.command().create(anyCmd));

    var query = requestFactory.query().all('type.spine.io/spine.web.test.given.Task');
    // Send the query and wait for all the results.
    var tasks = await client.fetch(query, Task.getDefault()).toList();
    print(tasks);
}
```
### API status

Currently, the Dart client is still under development. The formal version is `0.1.0`, however, the package is not actually published. So, it is both not recommended and technically challenging to use the client at this point.

### Further improvements
 - Simplify the API with the help of code generation.
 - Add support for entity/event subscriptions.
 - Add extension points to enable the client to work with non-Firebase Spine web APIs.

# Other changes

 - `firebase-server` package is removed from dependencies and the Firebase emulator is no longer started upon each build.
 - A servlet filter for CORS requests is added to the `test-app`.
 - For consistency, `web-tests` module which contained integration tests for the `client-js` lib is renamed into `js-tests`. Integration tests for the `client-dart` lib are placed into the `dart-test` module.

### CODE REVIEW

**Comment by dmdashenkov** [comment]
@armiol, PTAL.

File: client-dart/lib/uuids.dart [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+import 'package:uuid/uuid.dart';
+
+/// Creates a new UUID of version 4.
+///
+/// If [prefix] is specified, the result string starts with the given prefix.
```
Comment (armiol) [inline client-dart/lib/uuids.dart]: Let's add an empty `///`.

File: client-dart/test/time_test.dart [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+import 'package:spine_client/proto/main/dart/google/protobuf/timestamp.pb.dart';
+import 'package:spine_client/time.dart';
+import 'package:test/test.dart';
+
+void main() {
+    group('Time utils should', () {
```
Comment (armiol) [inline client-dart/test/time_test.dart]: Time utility should`?

File: integration-tests/test-app/src/main/java/io/spine/web/test/given/CrossOriginResourceSharingFilter.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+import static io.spine.web.CorsFilter.ANY_URL;
+import static io.spine.web.CorsFilter.NAME;
+
+/**
+ * A filter which allows requests from any origin to interact with any endpoint of this server.
+ *
+ * @see CorsFilter
+ */
+@WebFilter(filterName = NAME, urlPatterns = ANY_URL)
+public final class CrossOriginResourceSharingFilter extends CorsFilter {
```
Comment (armiol) [inline integration-tests/test-app/src/main/java/io/spine/web/test/given/CrossOriginResourceSharingFilter.java]: Can we rename it to `AnyOriginAllowedFilter`?

File: web/src/main/java/io/spine/web/CorsFilter.java [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+ * <p>Requests from any origin are allowed. The requests are allowed to contain credentials and
+ * the {@code Content-Type} header.
+ *
+ * <p>When configuring servlets via {@code web.xml}, users can add this filter to the requested
+ * servlets.
+ *
+ * <p>When configuring servlets via annotations, users can extend this class and mark the subtype
+ * with the {@link javax.servlet.annotation.WebFilter} annotation.
+ */
+public class CorsFilter implements Filter {
```
Comment (armiol) [inline web/src/main/java/io/spine/web/CorsFilter.java]: `AllowAllOriginsFilter`

**Pull request merged**